swift/aperture/main.swift

```diff
@@ -2,37 +2,51 @@ import Foundation
 import AVFoundation
 
 var recorder: Recorder!
+let arguments = CommandLine.arguments.dropFirst()
 
 func quit(_: Int32) {
   recorder.stop()
 }
 
-func record() throws {
-  // TODO: Use JSON and `Codable` here when Swift 4 is out
+struct CropArea: Decodable {
+  let x: CGFloat
+  let y: CGFloat
+  let width: CGFloat
+  let height: CGFloat
+}
+
+struct Options: Decodable {
+  let destination: String // TODO: Figure out a way to make this decodable into an `URL`
+  let fps: Int
+  let cropArea: CropArea? // TODO: Figure out a way to make this decodable into a `CGRect`
+  let showCursor: Bool
+  let highlightClicks: Bool
+  let displayId: String
+  let audioDeviceId: String?
+}
 
-  let args = CommandLine.arguments
-  let destination = args[1]
-  let fps = args[2]
-  let cropArea = args[3]
-  let showCursor = args[4]
-  let highlightClicks = args[5]
-  let displayId = args[6]
-  let audioDeviceId = args[7]
+func record() throws {
+  let json = arguments.first!.data(using: .utf8)!
+  let options = try JSONDecoder().decode(Options.self, from: json)
 
   var cropRect: CGRect?
-  if cropArea != "none" {
-    let points = cropArea.components(separatedBy: ":").map { Double($0)! }
-    cropRect = CGRect(x: points[0], y: points[1], width: points[2], height: points[3])
+  if let cropArea = options.cropArea {
+    cropRect = CGRect(
+      x: cropArea.x,
+      y: cropArea.y,
+      width: cropArea.width,
+      height: cropArea.height
+    )
   }
 
   recorder = try Recorder(
-    destination: URL(fileURLWithPath: destination),
-    fps: Int(fps)!,
+    destination: URL(fileURLWithPath: options.destination),
+    fps: options.fps,
     cropRect: cropRect,
-    showCursor: showCursor == "true",
-    highlightClicks: highlightClicks == "true",
-    displayId: displayId == "main" ? CGMainDisplayID() : CGDirectDisplayID(displayId)!,
-    audioDevice: audioDeviceId == "none" ? nil : .default(for: .audio)
+    showCursor: options.showCursor,
+    highlightClicks: options.highlightClicks,
+    displayId: options.displayId == "main" ? CGMainDisplayID() : CGDirectDisplayID(options.displayId)!,
+    audioDevice: options.audioDeviceId != nil ? AVCaptureDevice(uniqueID: options.audioDeviceId!) : .default(for: .audio)
```

Comment:
I keep wondering if it would be better to follow the Cocoa/UIKit style of optional arguments and assign after init instead.

Then we could do this instead. And we wouldn't have to repeat the default parameters each time.

```swift
recorder = try Recorder(…)

if let audioDeviceId = options.audioDeviceId {
  recorder.audioDevice = AVCaptureDevice(uniqueID: audioDeviceId)
}
```

I'm not really sure what the Swift best-practice is though? Maybe this way is more of a legacy Objective-C to Swift API quirk.

Reply:
Yeah, IMO assign after init would be better in this case 👍

Reply:
@LarsJK I'm confused about when you should prefer to assign after init and when you should have it in the constructor. Do you know what the best practice is? For example, should all optional values be assigned after init?

Reply:
Not sure what the best practice is for this, or if there is any..
The reason I feel it is better in this case is that you are duplicating the logic for the default value.